sdk/core/azure-core/src/base64.cpp

```diff
@@ -4,93 +4,474 @@
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1, // Bytes over 122 ('z') are invalid and cannot be decoded. Hence, padding the map with 255,
+        // which indicates invalid input
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+    -1,
+};
+
+static int32_t base64Encode(const uint8_t* threeBytes)
+{
+  int32_t i = (*threeBytes << 16) | (*(threeBytes + 1) << 8) | *(threeBytes + 2);
+
+  int32_t i0 = Base64EncodeArray[i >> 18];
+  int32_t i1 = Base64EncodeArray[(i >> 12) & 0x3F];
+  int32_t i2 = Base64EncodeArray[(i >> 6) & 0x3F];
+  int32_t i3 = Base64EncodeArray[i & 0x3F];
+
+  return i0 | (i1 << 8) | (i2 << 16) | (i3 << 24);
+}
+
+static int32_t base64EncodeAndPadOne(const uint8_t* twoBytes)
+{
+  int32_t i = (*twoBytes << 16) | (*(twoBytes + 1) << 8);
+
+  int32_t i0 = Base64EncodeArray[i >> 18];
+  int32_t i1 = Base64EncodeArray[(i >> 12) & 0x3F];
+  int32_t i2 = Base64EncodeArray[(i >> 6) & 0x3F];
+
+  return i0 | (i1 << 8) | (i2 << 16) | (EncodingPath << 24);
+}
+
+static int32_t base64EncodeAndPadTwo(const uint8_t* oneByte)
+{
+  int32_t i = (*oneByte << 8);
+
+  int32_t i0 = Base64EncodeArray[i >> 10];
+  int32_t i1 = Base64EncodeArray[(i >> 4) & 0x3F];
+
+  return i0 | (i1 << 8) | (EncodingPath << 16) | (EncodingPath << 24);
+}
+
+static void base64WriteIntAsFourBytes(char* destination, int32_t value)
+{
+  *(destination + 3) = static_cast<uint8_t>((value >> 24) & 0xFF);
+  *(destination + 2) = static_cast<uint8_t>((value >> 16) & 0xFF);
+  *(destination + 1) = static_cast<uint8_t>((value >> 8) & 0xFF);
+  *(destination + 0) = static_cast<uint8_t>(value & 0xFF);
+}
+
+std::string base64Encode(const std::vector<uint8_t>& data)
+{
+  int64_t sourceIndex = 0;
+  auto inputSize = data.size();
+  int32_t result = 0;
+  auto maxEncodedSize = ((inputSize + 2) / 3) * 4;
+  // Use a string with size to the max possible result
+  std::string encodedResult(maxEncodedSize, '0');
+  // Removing const from the string to update the placeholder string
+  auto destination = const_cast<char*>(encodedResult.data());
```

Comment:
I wonder if we should use some sort of string builder like stringstream instead of doing this.

@antkmsft, @CaseyCarter - what are your thoughts here?